### PR TITLE
#130 Possibility to share Reqwest client with Sentry through `transport` config.

### PR DESCRIPTION
**Pull request by mixalturek**
- Sharing through Arc is implemented only for Reqwest client that can be shared among threads.
- Curl client is not thread safe so it would require a mutex.

```
sentry::ClientOptions {
    transport: Box::new(|options: &sentry::ClientOptions| -> Box<sentry::internals::Transport> {
        Box::new(ReqwestHttpTransport::new(options, Some(http_client)))
    }),

    // ...
}
```

### CODE REVIEW

**Comment by mixalturek** [comment]
There are two implementations and pull requests for this feature. Please read the following first.
https://github.com/getsentry/sentry-rust/issues/130#issuecomment-468278410

File: src/transport.rs [hunk 1]
```diff
@@ -216,19 +221,10 @@ implement_http_transport! {
         signal: Arc<Condvar>,
         shutdown_immediately: Arc<AtomicBool>,
         queue_size: Arc<Mutex<usize>>,
+        http_client: Arc<Client>,
```
Comment (vorner) [inline src/transport.rs]: Just a random note… why `Arc`? `Client` is cheap to clone and shares the background thread, so there's no real reason to add another layer of indirection.

Comment (mixalturek) [reply src/transport.rs]: Right, it isn't necessary with this trick, thanks.

```
sentry::ClientOptions {
        transport: Box::new(|options: &sentry::ClientOptions| -> Box<sentry::internals::Transport> {
            Box::new(ReqwestHttpTransport::new(options, Some(reqwest::Client::clone(&arc_client))))
        }),
        // ...
}
```

File: src/transport.rs [hunk 2]
```diff
@@ -139,20 +140,24 @@ macro_rules! implement_http_transport {
 
         impl $typename {
             /// Creates a new transport.
-            pub fn new(options: &ClientOptions) -> $typename {
+            pub fn new(options: &ClientOptions, $hc_client: Option<$hc_client_ty>) -> $typename {
```
Comment (mitsuhiko) [inline src/transport.rs]: I would prefer if we add a `with_client` constructor that takes a client and leave `new` as it. That way upgrading is easier and also makes for a slightly nicer api. They can both call into an internal helper method.

Comment (mixalturek) [reply src/transport.rs]: Ok, I will update the code.

**Comment by mixalturek** [comment]
The failures are caused by recent release of stable Rust that made `trim_left_matches()` deprecated.

**Comment by mitsuhiko** [comment]
Thanks!

**Pull request merged**